Replace EFI boot entry during conversion

matejmatuska opened the pull request:
After an upgrade+conversion we want the EFI boot entry to have the name of the target OS (e.g. Red Hat Enterprise Linux, when converting from CS->RHEL). The source distro boot entry and also EFI directory (e.g. `/boot/efi/EFI/centos/`) should be removed.

This PR:
- splits out the functions from `common.grub` library that are not related to grub to 2 new libraries:
  - `common.efi` - functions related to UEFI; functions raise `EFIError` exception on errors instead
  - `common.partitons` (name suggestions are) welcome - operations on disks/parititions; this is used by both efi and grub lib; raise `StorageScanError` on errors
- functions from the `common.grub` library that are replaced by new libraries have been marked deprecated  
- refactor existing efi functions in the addarmbootloaderworkaround and removeupgradebootentry actors into the new efi lib
- adds a new actor that:
  - Removes the EFI dir of the source distro (e.g. `/boot/efi/EFI/centos/`).
    - If not empty create a post upgrade report to handle the leftover files manually.
  - Removes the EFI boot entry of the source distro.
  - Adds a new EFI boot entry for the target distro.
    - The contents of the target distro EFI dir are brought in by the grub-efi-x64 and shim-x64 packages.
- adds another actor that inhibits conversions on systems with secure boot enabled
  - At least for now, I am going with an inhibitor as this current solution doesn't work with secureboot enabled. The problem probably is, in my understanding, that the shim and grub only contain keys for verifying signatures of the source distro. Our upgrade boot entry uses the target distros kernel, which is signed by the target distro. And so the shim/grub of the source distro rejects it.
  - **A possible solution** IMO is to create an EFI boot entry for the upgrade pointing to the target distro's shim, similarly to how it's done in the addarmbootloaderworkaround actor.

Jira: RHEL-133549

### TODO
- [x] ARM - should work
- [X] secure boot - For now inhibited, see explanation above.
- [x] secure boot inhibitor - Need to get a better link for the report.

matejmatuska committed:
Split grub lib into multiple dedicated libs

The grub library has been used for all things related to booting,
including also utilities related to EFI and "partitions".

This patch splits the unrelated information grub library into multiple
dedicated libraries:
- grub - now only grub related utilities
- firmware - contains the is_efi and is_bios helpers
- firmware/efi - utilities related to EFI
- partitions - contains utilities related to partitions and devices, used by
               both grub and efi

Functions in efi library now raise efi.EFIError instead of
StopActorExecutionError and StopActorExecution

Functions in the partitions library raises partitions.StorageScanError
exception instead of StopActorExecution.

matejmatuska committed:
Inhibit conversions with secure boot enabled

The EFI bootloader entry replacement during conversion, handled by the
convert/updateefi actor, does currently not handle secure boot. This
patch adds the convert/securebootinbihit actor that inhibits secureboot
during conversion, regular upgrade is not affected.

Jira: RHEL-133549

matejmatuska committed:
Update EFI entry and directory during conversion

After an upgrade+conversion, the upgraded system might end up with an
EFI boot loader entry from the source system pointing the shim or grub
EFI binary from the source system. E.g. Centos Stream pointing to
\EFI\centos\shimx64.efi instead of Red Hat Enterprise Linux pointing to
\EFI\redhat\shimx64.efi.

This patch introduces a new convert/updateefi actor that adds a new EFI
boot entry for the target system with the correct EFI binary, removes
the source system directory on ESP, e.g /boot/efi/EFI/centos, and
finally removes the source system EFI boot entry (see below).

This is handled for both x86_64 and aarch64 EFI systems. However,
in case of IPU 8 -> 9 on ARM systems it's little bit conflicting with
the existing fixes (the efi_reboot_fix library) which expects already
to set the EFI boot entry for the next boot. We will need to look
at it closer in future but partially it's resolved already.

Additionally, if the source EFI directory is not empty (usually when
it contains custom files), do not remove it and produce a post-upgrade
report instead, instructing the user to review the content and handle
it manually.

This currently does not work with SecureBoot enabled. The problem
probably is, in my understanding, that the shim and grub only contain
keys for verifying signatures of the source distro. Our upgrade boot
entry uses the target distros kernel, which is signed by the target
distro. And so the shim/grub of the source distro rejects it.

Jira: RHEL-133549

danmyway commented:
/packit build

pirat89 commented on repos/system_upgrade/common/actors/convert/securebootinbihit/libraries/securebootinhibit.py:
```diff
@@ -0,0 +1,30 @@
+            reporting.Summary((
+                'Upgrades with conversions are currently not possible when SecureBoot'
+                ' is enabled in the UEFI.'
+            )),
```
```suggestion
            reporting.Title('Detected enabled Secure Boot when trying to convert the system'),
            reporting.Summary((
                'The conversion to a different linux distribution is not possible when the Secure Boot'
                ' is enabled. Artifacts in the target linux distribution use are signed by different keys'
                ' that are not accepted by current linux distribution.'
            )),
```

matejmatuska replied:
Thanks, the report was still a little bit WIP.

pirat89 reviewed (changes requested):
my apology for partial review - maybe is valid, I haven't finished the full review yet so I am still considering some stuff. But sharing as for sure some parts will need to be made differently.

pirat89 commented on repos/system_upgrade/common/actors/convert/securebootinbihit/libraries/securebootinhibit.py:
@matejmatuska  btw, typo in dirname: securebootin**bihit** -> securebootin**hibit**

pirat89 commented on repos/system_upgrade/common/libraries/firmware/efi.py:
```diff
@@ -0,0 +1,325 @@
+        "--loader",
+        efi_bin_path,
+        "--label",
+        label,
```
nitpick: I will be old school, and say that this is for me more readable :) but I am ok with original formatting as well.
```suggestion
        "/usr/sbin/efibootmgr",
        "--create",
        "--disk", efi_dev,
        "--part", str(esp_number),
        "--loader", efi_bin_path,
        "--label", label,
```

pirat89 commented on repos/system_upgrade/common/libraries/distro.py:
```diff
@@ -208,3 +208,16 @@ def get_distro_repoids(context, distro, major_version, arch):
+    The pretty name is what is found in the NAME field of /etc/os-release.
+    """
+    return {
+        "rhel": "Red Hat Enterprise Linux",
```
didn't we agree to use just `"RHEL` acronym actually? I dunno honestly, but as we usually do not use the full name in reports, I would expect to use `RHEL` instead.

pirat89 replied:
Just thinking whether it will not be a little bit confusing with anothe rmap we plan to deliver. But I possibly overthinking it right now. Let's discuss this one later.

pirat89 replied:
it's possible :D

pirat89 commented on repos/system_upgrade/common/actors/systemfacts/libraries/systemfacts.py:
```diff
@@ -297,10 +297,26 @@ def _get_firewall_status(service_name):
-    return FirmwareFacts(firmware=firmware)
+        ppc64le_opal = os.path.isdir('/sys/firmware/opal/')
+
+    is_secureboot = None
```
@matejmatuska  why not make it False by default already when allowed value is boolean?

matejmatuska replied:
I replied to the comment in the model.

pirat89 commented on repos/system_upgrade/common/libraries/grub.py:
```diff
@@ -16,20 +17,26 @@
+@deprecated(
+    since="2025-12-04",
+    message="Replaced by canonical_path_to_efi_format in the firmware.efi library",
+)
```
do not forget to update docs about all deprecations

matejmatuska replied:
Thanks, I would.

pirat89 commented on repos/system_upgrade/common/models/firmwarefacts.py:
```diff
@@ -9,3 +9,5 @@ class FirmwareFacts(Model):
     ppc64le_opal = fields.Nullable(fields.Boolean())
     """ Check OPAL presence to identify ppc64le bare metal systems """
+    secureboot_enabled = fields.Nullable(fields.Boolean())
+    """ Whether SecureBoot is enabled, always False on BIOS systems """
```
do we expect anything than True/False to make it nullable?

matejmatuska replied:
Not really, but same for the `ppc64le_opal` and yet it's set to `None` when not on ppcle. So it means sonmething like "not applicable".
I did the same with the secureboot which is "not applicable" i.e. `None` on BIOS systems.

pirat89 commented on repos/system_upgrade/el8toel9/actors/addarmbootloaderworkaround/libraries/addupgradebootloader.py:
```diff
@@ -19,9 +13,8 @@
-LEAPP_EFIDIR_CANONICAL_PATH = os.path.join(EFI_MOUNTPOINT, 'EFI/leapp/')
-RHEL_EFIDIR_CANONICAL_PATH = os.path.join(EFI_MOUNTPOINT, 'EFI/redhat/')
+LEAPP_EFIDIR_CANONICAL_PATH = os.path.join(efi.EFI_MOUNTPOINT, 'EFI/leapp/')
+RHEL_EFIDIR_CANONICAL_PATH = os.path.join(efi.EFI_MOUNTPOINT, 'EFI/redhat/')
```
@matejmatuska  Hmm, in this way, it would be working just for ARM on RHEL. Shouldn't it be dynamic? having a function in efi lib saying what is the expected efidir path? Also it could be broken in case of conversion. and now I do not remember architectures of other distros. so pointing this out at least

matejmatuska replied:
Right, definitely good point about the distro name.

matejmatuska commented on repos/system_upgrade/common/actors/convert/securebootinbihit/tests/test_securebootinhibit.py:
```diff
@@ -0,0 +1,58 @@
+            FirmwareFacts(firmware='bios', ppc64le_opal=None, secureboot_enabled=True),
+            True,
+            False
+        ),
```
good catch

matejmatuska commented on repos/system_upgrade/common/libraries/firmware/efi.py:
```diff
@@ -0,0 +1,325 @@
+        raise EFIError(
+            "Unable to add a new UEFI bootloader entry '{}' for EFI binary at {}.".format(
+                label, efi_bin_path
+            )
```
Thanks, I am still forgetting that ocasionaly.

matejmatuska commented on repos/system_upgrade/el8toel9/actors/addarmbootloaderworkaround/libraries/addupgradebootloader.py:
```diff
@@ -59,21 +52,31 @@ def process():
                 upgrade_entry=EFIBootEntry(**{f: getattr(upgrade_boot_entry, f) for f in efibootentry_fields}),
                 upgrade_bls_dir=UPGRADE_BLS_DIR,
-                upgrade_entry_efi_path=os.path.join(EFI_MOUNTPOINT, LEAPP_EFIDIR_CANONICAL_PATH),
+                upgrade_entry_efi_path=os.path.join(efi.EFI_MOUNTPOINT, LEAPP_EFIDIR_CANONICAL_PATH),
```
It is, good catch

matejmatuska commented on repos/system_upgrade/common/libraries/firmware/efi.py:
```diff
@@ -0,0 +1,325 @@
+
+
+def canonical_path_to_efi_format(canonical_path):
+    r"""Transform the canonical path to the UEFI format.
```
Ah nevermind the r has to be there for the backslashes.

matejmatuska commented on repos/system_upgrade/common/libraries/firmware/efi.py:
```diff
@@ -0,0 +1,325 @@
+
+
+# TODO maybe rename this to find_boot_entry
+def get_boot_entry(efibootinfo, label, efi_bin_path):
```
I kept it as it is

matejmatuska commented on repos/system_upgrade/common/actors/convert/updateefi/libraries/updateefi.py:
```diff
@@ -0,0 +1,191 @@
+    if distro_id == "rhel":
+        return os.path.join(efi.EFI_MOUNTPOINT, "EFI", "redhat")
+
+    return os.path.join(efi.EFI_MOUNTPOINT, "EFI", distro_id)
```
Okay, I will move it to the distro lib for now, but I think we should have some followup where we move it to some config file along with the other "static" information about distro. It would make enabling of other distros nicer.

matejmatuska committed:
systemfacts: Fix the scan on VMs with UEFI without Secure Boot support

In some VMs the Secure Boot is not supported even when UEFI is used.
On such system the scanner has been failing constantly. Check the
error properly and set secure boot to None in such a case.

matejmatuska committed:
addarmbootloaderworkaround: Make the actor distro agnostic

The path to the EFI dir has been hardcoded just for RHEL. Obtain
the expected canonical path to the default EFI boot entry dir
from `leapp.libraries.common.distro.get_distro_efidir_canon_path`
instead.

pirat89 committed:
grub lib: Rigidify: Raise proper exceptions on errors

Previously, number of functions raised StopActorExecution on errors
or ignored them, deciding the logic inside the library instead of in
actors. This could potentially lead to hidden problems and bugs for
which we do not have a good explanation due to missing data. This
has been happening mainly for `get_grub_devices()`.

* `has_grub()` - raises OSError if cannot read from a given block device
* `get_grub_devices()` - raises grub.GRUBDeviceError when cannot obtain
  devices hosting GRUB

Also update actors calling `get_grub_devices()` to properly handle
the situation.

matejmatuska committed:
lib/efi: Refactor EFI related functions into the efi shared lib

pirat89 committed:
distro lib: introduce get_distro_efidir_canon_path()

The function returns expected (default) canonical path for the default
efi directory (the directory under /boot/efi/EFI/ where the usual
EFI boot entry is supposed to be located) based on the specified
linux distribution.

E.g. for RHEL: /boot/efi/EFI/redhat

The path might be incorrect on other distributions that are not covered
correctly yet. In such a case, the default is:
  /boot/efi/EFI/redhat/<distro_id>

pirat89 reviewed (approved):
lgtm and it seems to be working - however we have not been able to check all calls from actors yet to potentially catch all possible exceptions. Due to high amount of changes, we decided to merge it in this current state and cover remaining possible overlooked calls in a followup PR later this week.
